File: RobotSimulatorC#/RobotSimulatorTestApp/IRotation.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace RobotSimulatorTestApp
{
    public interface IRotation
    {
        Direction Rotate(Direction direction);
    }

    public class RotateRight : IRotation
    {
        public Direction Rotate(Direction direction)
        {
            return (Direction)(((int)direction + 1) % 4);
        }
    }

    public class RotateLeft : IRotation
    {
        public Direction Rotate(Direction direction)
        {
            return (Direction)(((int)direction + 3) % 4);
        }
    }
}

File: 1_RobotSimulatorJava/IRotation.java
// Design pattern: Factory method
public interface IRotation {
    Direction rotate(Direction direction);
}

class RotateRight implements IRotation {
    @Override
    public Direction rotate(Direction direction) {
        return Direction.values()[(direction.ordinal() + 1) % 4];
    }
}

class RotateLeft implements IRotation {
    @Override
    public Direction rotate(Direction direction) {
        return Direction.values()[(direction.ordinal() + 3) % 4];
    }
}
